extension/goplugin/util.go

```diff
@@ -74,58 +73,157 @@ func Finish(testReporter gomock.TestReporter) {
 		}
 		kind := fieldType.Type.Kind()
-		if strings.Contains(fieldType.Type.String(), "goext.Null") {
-			if context[propertyName] == nil {
-				field.FieldByName("Valid").SetBool(false)
-			} else {
-				field.FieldByName("Valid").SetBool(true)
-				value := reflect.ValueOf(context[propertyName])
-				field.FieldByName("Value").Set(value)
+		mapValue, mapValueExists := context[propertyName]
+
+		if kind == reflect.Interface {
+			if field.IsNil() && mapValue == nil {
+				continue
 			}
-		} else if kind == reflect.Struct || kind == reflect.Ptr || kind == reflect.Slice {
-			mapJSON, err := json.Marshal(context[propertyName])
-			if err != nil {
-				return nil, err
+			field.Set(reflect.ValueOf(mapValue))
+		} else if isMaybeType(kind, fieldType.Type.String()) {
+			if kind == reflect.Ptr {
+				if !mapValueExists {
+					continue
+				}
+				field.Set(reflect.New(fieldType.Type.Elem()))
+				field = field.Elem()
 			}
-			newField := reflect.New(field.Type())
-			fieldJSON := string(mapJSON)
-			fieldInterface := newField.Interface()
-			err = json.Unmarshal([]byte(fieldJSON), &fieldInterface)
-			if err != nil {
-				return nil, err
+			if err := maybeFromMap(context, propertyName, field); err != nil {
+				return err
+			}
+		} else if kind == reflect.Struct || kind == reflect.Ptr {
+			if mapValue != nil {
+				field.Set(reflect.ValueOf(reflect.New(field.Type()).Elem().Interface()))
+				if err := resourceFromMap(mapValue.(map[string]interface{}), field); err != nil {
+					return err
+				}
+			}
+		} else if kind == reflect.Slice {
+			if err := sliceToMap(context, propertyName, field); err != nil {
+				return err
 			}
-			field.Set(newField.Elem())
 		} else {
-			value := reflect.ValueOf(context[propertyName])
-			if value.IsValid() {
-				if value.Type() == field.Type() {
-					field.Set(value)
-				} else {
-					if field.Kind() == reflect.Int && value.Kind() == reflect.Float64 { // reflect treats number(N, 0) as float
-						field.SetInt(int64(value.Float()))
-					} else {
-						return nil, fmt.Errorf("invalid type of '%s' field (%s, expecting %s)", propertyName, value.Kind(), field.Kind())
+			if err := assignMapValueToField(mapValue, propertyName, field); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func assignMapValueToField(mapValue interface{}, fieldName string, field reflect.Value) error {
+	value := reflect.ValueOf(mapValue)
+	if value.IsValid() {
+		if value.Type() == field.Type() {
+			field.Set(value)
+		} else {
+			if field.Kind() == reflect.Int && value.Kind() == reflect.Float64 { // reflect treats number(N, 0) as float
+				field.SetInt(int64(value.Float()))
+			} else {
+				return fmt.Errorf("invalid type of '%s' field (%s, expecting %s)", fieldName, value.Kind(), field.Kind())
+			}
+		}
+	}
+	return nil
+}
+
+func isMaybeType(kind reflect.Kind, typeName string) bool {
+	return anyOfKinds(kind, []reflect.Kind{reflect.Ptr, reflect.Struct}) && strings.HasPrefix(typeName, "goext.Maybe")
```

Comment:
I don't like checking substrings because this is an operation that compiler cannot optimize. Maybe it is possible to add simple comma-separated switch here?

Reply:
I thought there's not but actually it's possible to use switch. Thanks